[6.2][AST/Sema] SE-0487: Implement `@nonexhaustive` attribute and new enum exhaustivity checking rule

xedin opened the pull request:
- Explanation:

  This is an implementation of a second revision of SE-0487 proposal to
  make non-frozen enums extensible for non-resilient cross-module access.

  The new attribute (`@nonexhaustive`) controls whether cross-module access 
  to the declaration needs `@unknown default:` because it's allowed to gain new
  cases even if the module is non-resilient. A variant of the flag `@nonexhaustive(warn)`
  could be used by API authors to push error about non-exhaustive use farther into the 
  future.

- Resolves: rdar://154965992

- Main Branch PRs: https://github.com/swiftlang/swift/pull/80503, https://github.com/swiftlang/swift/pull/81073, https://github.com/swiftlang/swift/pull/81176, https://github.com/swiftlang/swift/pull/82807

- Risk: Very Low. These changes are guarded by an experimental feature `NonexhaustiveAttribute` which is not enabled by default yet.
 
- Reviewed By: @FranzBusch    

- Testing: Added new test-cases.


<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST/Sema] Add `@extensible` attribute on `enum` declarations

This attribute controls whether cross-module access to the declaration
needs `@unknown default:` because it's allowed to gain new cases even
if the module is non-resilient.

(cherry picked from commit a0ae93d3a8aa84dda4e3489e88b4482215e43821)

xedin committed:
[Frontend/Serialization] Remove `ExtensibleEnums` experimental flag

For now the semantics provided by `@extensible` keyword on per-enum
basis. We might return this as an upcoming feature in the future with
a way to opt-out.

(cherry picked from commit bf19481ab6b3bc80ed50ba243ba5c4215a01d4b4)

xedin committed:
[Frontend] Add `ExtensibleAttribute` to guard use of `@extensible` attribute

Guard against condfails when older compilers get a swift interface
that uses `@extensible` attribute. The attribute itself doesn't
have any effect in swift interfaces yet since all of the public
enums are already resilient in that mode.

(cherry picked from commit 6d89bca76501812a8813bd79a8e48b6bc1debd4a)

xedin committed:
[AST] Prevent use of `@extensible` without `ExtensibleAttribute` feature

(cherry picked from commit 0ad7d8b59096a6888499c3415b9a0f48cd32f865)

xedin committed:
[AST/Sema] Add `@preEnumExtensibility` attribute to downgrade warnings for extensible enums

Just like `@preconcurrency` for concurrency, this attribute is going
to allow exhaustiveness error downgrades for enums that were retroactively
marked as `@extensible`.

(cherry picked from commit 498430afafa633b2e3f7a4ab546b8caf901803a3)

xedin committed:
[AST/Sema] SE-0487: Rename `@extensible` into `@nonexhaustive`

This is an accepted spelling for the attribute. This commit
also renames the feature flag from `ExtensibleAttribute` to
`NonexhaustiveAttribute` to match the spelling of the attribute.

(cherry picked from commit fe1ae7571197d67afe06032c9788a0cc097db9d5)

xedin committed:
[AST/Sema] SE-0487: Expand `@nonexhaustive` attribute to support `warn` argument

The spelling `@nonexhaustive(warn)` replaces `@preEnumExtensibility`
attriubte.

(cherry picked from commit 43eec8fede049e3073748be421fcf4207c1e4135)

xedin committed:
[AST/Sema] SE-0487: Remove `@preEnumExtensibility` attribute

(cherry picked from commit 0444d297b6fddcd5cd3adbc7ccd34e120ef3ebf8)

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Windows platform

FranzBusch commented on include/swift/Basic/Features.def:
```diff
@@ -525,6 +520,9 @@ EXPERIMENTAL_FEATURE(AllowRuntimeSymbolDeclarations, true)
 EXPERIMENTAL_FEATURE(CopyBlockOptimization, true)
 
+/// Allow use of `@nonexhaustive` on public enums
+SUPPRESSIBLE_EXPERIMENTAL_FEATURE(NonexhaustiveAttribute, false)
```
Should this be an upcoming feature at this point?

FranzBusch replied:
Yes and [the proposal](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0487-extensible-enums.md) says `Upcoming Feature Flag: ExtensibleAttribute`

xedin replied:
Yes indeed. I am going to open a PR to update that on main.

xedin replied:
✅

xedin committed:
[AST] SE-0487: Promote `NonexhaustiveAttribute` to a language feature

The proposal [has been accepted](https://forums.swift.org/t/accepted-se-0487-nonexhaustive-enums/81508)
which makes the feature experimental no longer.

(cherry picked from commit a7ecd309d4577b7974a2c912db3f3b23f83a61be)

xedin commented:
@swift-ci please test
